src/extract/faster_bottom_up.rs

```diff
@@ -0,0 +1,117 @@
+use super::*;
+
+/// A faster bottom up extractor inspired by the faster-greedy-dag extractor.
+pub struct BottomUpExtractor;
+
+impl Extractor for BottomUpExtractor {
+    fn extract(&self, egraph: &EGraph, _roots: &[ClassId]) -> ExtractionResult {
+        // 1. build map from class to parent nodes
+        let mut parents = IndexMap::<ClassId, Vec<NodeId>>::default();
+        let n2c = |nid: &NodeId| egraph.nid_to_cid(nid);
+
+        for class in egraph.classes().values() {
+            parents.insert(class.id.clone(), Vec::new());
+        }
+        for class in egraph.classes().values() {
+            for node in &class.nodes {
+                for c in &egraph[node].children {
+                    parents[n2c(c)].push(node.clone());
+                }
+            }
+        }
+
+        // 2. start analysis from leaves
+        let mut analysis_pending = UniqueQueue::default();
+
+        for class in egraph.classes().values() {
+            for node in &class.nodes {
+                if egraph[node].is_leaf() {
+                    analysis_pending.insert(node.clone());
+                }
+            }
+        }
+
+        let mut result = ExtractionResult::default();
+        let mut costs = IndexMap::<ClassId, Cost>::default();
+
+        while let Some(node_id) = analysis_pending.pop() {
+            let class_id = n2c(&node_id);
+            let node = &egraph[&node_id];
+            if node.children.iter().all(|c| costs.contains_key(n2c(c))) {
+                let prev_cost = costs.get(class_id).unwrap_or(&INFINITY);
+
+                let cost = result.node_sum_cost(egraph, node, &costs);
+                if cost < *prev_cost {
+                    result.choose(class_id.clone(), node_id.clone());
+                    costs.insert(class_id.clone(), cost);
+                    analysis_pending.extend(parents[class_id].iter().cloned());
+                }
+            } else {
+                analysis_pending.insert(node_id.clone());
+            }
+        }
+
+        result
+    }
+}
+
+/** A data structure to maintain a queue of unique elements.
```

Comment:
Why copy-paste this from the other file? Can we move it to a shared module?

Reply:
I'm against sharing code between these algorithms for now: I think it's good to be able to tweak them without worrying about changing the comparison.